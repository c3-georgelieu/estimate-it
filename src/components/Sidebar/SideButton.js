import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import FA from 'react-fontawesome'
import { Button } from 'reactstrap'

export const SideButton = ({
  title,
  name,
  color,
  onClick,
  link,
}) => (
  <div className="side-button">
    {link ? (
      <Link to={link}>
        <Button
          outline
          color={color}
        >
          <FA name={name} />
        </Button>
      </Link>
    ) : (
      <Button
        outline
        color={color}
        onClick={onClick}
      >
        <FA name={name} />
      </Button>
    )}
    <div className={`title text-color-${color}`}>
      {title}
    </div>
  </div>
)

SideButton.propTypes = {
  title: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  color: PropTypes.string,
  onClick: PropTypes.func,
  link: PropTypes.string,
}

SideButton.defaultProps = {
  color: 'primary',
  onClick: undefined,
  link: undefined,
}

export default SideButton
