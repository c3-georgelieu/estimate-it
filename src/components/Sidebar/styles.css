.estimate aside {
  width: 72px;
  --offset: -72px;
  padding-top: 10px;
  background-color: #222;
  color: #fff;
  margin: 0;
  display: flex;
  flex-direction: column;
  overflow-y: scroll;
}

.sidebar-padding {
  flex-grow: 1;
}
