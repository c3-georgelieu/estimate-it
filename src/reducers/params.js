import { SET_PARAMS } from '../actions/types'

const defaultState = {
  apiKey: '',
  estimateId: '',
}

export default (state = defaultState, { type, payload }) => {
  switch (type) {
    case SET_PARAMS:
      return ({
        ...state,
        apiKey: payload.apiKey,
        estimateId: payload.estimateId,
      })

    default:
      return state
  }
}
