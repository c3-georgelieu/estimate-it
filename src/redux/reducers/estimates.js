import { handleText } from '../../helpers/task-parsing'
import {
  ADD_ESTIMATE,
  RECALCULATE,
  CLEAN_ESTIMATE,
} from '../actions/types'

const emptyEstimate = ({
  _id: 'new',
  text: '',
  graphData: [],
  project: 'New Project',
})

const defaultState = {
  new: emptyEstimate,
}

export default (state = defaultState, { type, payload }) => {
  switch (type) {
    case ADD_ESTIMATE:
      return ({
        ...state,
        [payload._id]: { ...payload },
      })
    case CLEAN_ESTIMATE:
      return ({
        ...state,
        // Put 'new' here to avoid creating an estimate with an undefined `_id`
        // ...while logging out on different routes
        // ...(that may not contain an `:estimateId` param)
        [payload.estimateId || 'new']: emptyEstimate,
      })
    case RECALCULATE:
      return ({
        ...state,
        [payload._id]: {
          _id: payload._id,
          ...handleText(state[payload._id].text),
        },
      })

    default:
      return state
  }
}
