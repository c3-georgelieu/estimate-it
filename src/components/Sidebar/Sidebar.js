import React from 'react'
import PropTypes from 'prop-types'
import SideButton from '../SideButton'
import { estimateType } from '../Estimate/propTypes'

const Sidebar = ({
  match: { params: { estimateId } },
  recalc,
  saveEstimate,
  openAuthScreen,
  estimates,
  username,
  graphView,
  minifyGraph,
  enlargeGraph,
}) => {
  const { saved, calculated } = estimates[estimateId]
  return (
    <aside>
      <SideButton
        title="New"
        name="file"
        disabled={estimateId === 'new'}
        link="/estimate/new"
      />
      <SideButton
        title="Erase"
        name="eraser"
        color="danger"
      />
      <SideButton
        title="Calculate"
        name="rocket"
        disabled={calculated}
        onClick={() => !calculated && recalc(estimateId)}
      />
      <SideButton
        title="Save"
        name="save"
        disabled={!username || saved}
        color={(saved && 'secondary') || (username && calculated && 'primary') || 'warning'}
        onClick={() => !saved && saveEstimate({ estimateId })}
      />
      {graphView === 'minified' ? (
        <SideButton
          title="Graph"
          name="bar-chart"
          onClick={enlargeGraph}
        />
      ) : (
        <SideButton
          title="Editor"
          name="edit"
          onClick={minifyGraph}
        />
      )}
      <SideButton
        title="Dashboard"
        name="list-ul"
        disabled={!username}
        link="/dashboard"
      />
      <div className="sidebar-padding" />
      <SideButton
        title="Auth"
        name="key"
        color={username ? 'success' : 'warning'}
        onClick={openAuthScreen}
      />
      <SideButton
        title="Home"
        name="home"
        link="/"
      />
    </aside>
  )
}

Sidebar.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      estimateId: PropTypes.string,
    }).isRequired,
  }).isRequired,
  recalc: PropTypes.func.isRequired,
  saveEstimate: PropTypes.func.isRequired,
  openAuthScreen: PropTypes.func.isRequired,
  estimates: PropTypes.objectOf(estimateType).isRequired,
  username: PropTypes.string,
  graphView: PropTypes.string.isRequired,
  minifyGraph: PropTypes.func.isRequired,
  enlargeGraph: PropTypes.func.isRequired,
}

Sidebar.defaultProps = {
  username: undefined,
}

export default Sidebar
