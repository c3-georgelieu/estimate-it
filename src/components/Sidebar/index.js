import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import Sidebar from './Sidebar'
import './styles.css'
import {
  recalc,
  openAuthScreen,
  closeAuthScreen,
} from '../../redux/actions'
import {
  saveEstimate,
} from '../../redux/actions/async'

const mapStateToProps = ({ estimates }) => ({
  estimates,
})

const mapDispatchToProps = ({
  recalc,
  saveEstimate,
  openAuthScreen,
  closeAuthScreen,
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Sidebar))
