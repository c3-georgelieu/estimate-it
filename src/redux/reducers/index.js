import { combineReducers } from 'redux'
import { reducer as toastrReducer } from 'react-redux-toastr'
import estimates from './estimates'
import visualEffects from './visualEffects'
import redirector from './redirector'
import creds from './creds'

export default combineReducers({
  estimates,
  visualEffects,
  toastr: toastrReducer,
  redirector,
  creds,
})
