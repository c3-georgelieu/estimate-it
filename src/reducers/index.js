import { combineReducers } from 'redux'
import estimate from './estimate'

export default combineReducers({
  estimate,
})
