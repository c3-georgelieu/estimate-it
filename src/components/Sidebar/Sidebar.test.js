import React from 'react'
import { configure, mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Sidebar from './Sidebar'
import { emptyEstimate } from '../../redux/reducers/estimates'

it('renders correctly', () => {
  configure({ adapter: new Adapter() })
  const tree = shallow(<Sidebar
    match={{ params: { estimateId: 'new' } }}
    recalc={_ => null}
    saveEstimate={_ => null}
    openAuthScreen={_ => null}
    estimates={{ new: emptyEstimate }}
    username="Test User"
    graphView="minified"
    minifyGraph={_ => null}
    enlargeGraph={_ => null}
  />, {})
  expect(tree).toMatchSnapshot()
})
