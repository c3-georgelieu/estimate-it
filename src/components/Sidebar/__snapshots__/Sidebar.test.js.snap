// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Sidebar
    enlargeGraph={[Function]}
    estimates={
        Object {
            "new": Object {
              "_id": "new",
              "calculated": true,
              "graphData": Array [],
              "participants": Array [],
              "project": "",
              "saved": false,
              "text": "",
            },
          }
    }
    graphView="minified"
    match={
        Object {
            "params": Object {
              "estimateId": "new",
            },
          }
    }
    minifyGraph={[Function]}
    openAuthScreen={[Function]}
    recalc={[Function]}
    saveEstimate={[Function]}
    username="Test User"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(SideButton)
          disabled={true}
          link="/estimate/new"
          name="file"
          title="New"
/>,
        <Connect(SideButton)
          color="danger"
          name="eraser"
          title="Erase"
/>,
        <Connect(SideButton)
          disabled={true}
          name="rocket"
          onClick={[Function]}
          title="Calculate"
/>,
        <Connect(SideButton)
          color="primary"
          disabled={false}
          name="save"
          onClick={[Function]}
          title="Save"
/>,
        <Connect(SideButton)
          name="bar-chart"
          onClick={[Function]}
          title="Graph"
/>,
        <Connect(SideButton)
          disabled={false}
          link="/dashboard"
          name="list-ul"
          title="Dashboard"
/>,
        <div
          className="sidebar-padding"
/>,
        <Connect(SideButton)
          color="success"
          name="key"
          onClick={[Function]}
          title="Auth"
/>,
        <Connect(SideButton)
          link="/"
          name="home"
          title="Home"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": true,
          "link": "/estimate/new",
          "name": "file",
          "title": "New",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "color": "danger",
          "name": "eraser",
          "title": "Erase",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": true,
          "name": "rocket",
          "onClick": [Function],
          "title": "Calculate",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "color": "primary",
          "disabled": false,
          "name": "save",
          "onClick": [Function],
          "title": "Save",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "name": "bar-chart",
          "onClick": [Function],
          "title": "Graph",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "link": "/dashboard",
          "name": "list-ul",
          "title": "Dashboard",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "sidebar-padding",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "color": "success",
          "name": "key",
          "onClick": [Function],
          "title": "Auth",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "link": "/",
          "name": "home",
          "title": "Home",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "aside",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(SideButton)
            disabled={true}
            link="/estimate/new"
            name="file"
            title="New"
/>,
          <Connect(SideButton)
            color="danger"
            name="eraser"
            title="Erase"
/>,
          <Connect(SideButton)
            disabled={true}
            name="rocket"
            onClick={[Function]}
            title="Calculate"
/>,
          <Connect(SideButton)
            color="primary"
            disabled={false}
            name="save"
            onClick={[Function]}
            title="Save"
/>,
          <Connect(SideButton)
            name="bar-chart"
            onClick={[Function]}
            title="Graph"
/>,
          <Connect(SideButton)
            disabled={false}
            link="/dashboard"
            name="list-ul"
            title="Dashboard"
/>,
          <div
            className="sidebar-padding"
/>,
          <Connect(SideButton)
            color="success"
            name="key"
            onClick={[Function]}
            title="Auth"
/>,
          <Connect(SideButton)
            link="/"
            name="home"
            title="Home"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": true,
            "link": "/estimate/new",
            "name": "file",
            "title": "New",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "color": "danger",
            "name": "eraser",
            "title": "Erase",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": true,
            "name": "rocket",
            "onClick": [Function],
            "title": "Calculate",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "color": "primary",
            "disabled": false,
            "name": "save",
            "onClick": [Function],
            "title": "Save",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "name": "bar-chart",
            "onClick": [Function],
            "title": "Graph",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "link": "/dashboard",
            "name": "list-ul",
            "title": "Dashboard",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "sidebar-padding",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "color": "success",
            "name": "key",
            "onClick": [Function],
            "title": "Auth",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "link": "/",
            "name": "home",
            "title": "Home",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "aside",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
